File: stonex_ops/audit.py
"""Audit log: records who / when / what / result for every MCP tool call.

Format: JSONL (one line per entry), written to stderr and an optional file.
"""

from __future__ import annotations

import json
import sys
import threading
import uuid
from dataclasses import asdict, dataclass
from enum import Enum
from pathlib import Path
from typing import Any, Optional, TextIO


class AuditStatus(str, Enum):
    success = "success"
    execution_failed = "execution_failed"


@dataclass
class AuditEntry:
    timestamp: str
    tool: str
    arguments: dict[str, Any]
    result_status: AuditStatus
    error: Optional[str]
    duration_ms: int
    session_id: str


class AuditLogger:
    """Thread-safe JSONL audit logger."""

    def __init__(self, audit_file: Optional[Path] = None) -> None:
        self._session_id = str(uuid.uuid4())
        self._lock = threading.Lock()
        self._file: Optional[TextIO] = None
        if audit_file:
            self._file = open(audit_file, "a", encoding="utf-8")  # noqa: SIM115

    @property
    def session_id(self) -> str:
        return self._session_id

    def log(self, entry: AuditEntry) -> None:
        line = json.dumps(asdict(entry), ensure_ascii=False)
        with self._lock:
            if self._file:
                self._file.write(line + "\n")
                self._file.flush()
        # Always write to stderr for dev visibility.
        print(f"[audit] {line}", file=sys.stderr)

    def close(self) -> None:
        if self._file:
            self._file.close()
            self._file = None

File: src/audit.rs
//! 审计日志:记录每次 MCP tool 调用的 who / when / what / result。
//!
//! 日志格式: JSONL (每行一条),写入 stderr 或指定文件。

use serde::Serialize;
use std::path::Path;
use std::sync::Mutex;

#[derive(Debug, Clone, Serialize)]
pub(crate) struct AuditEntry {
    pub timestamp: String,
    pub tool: String,
    pub arguments: serde_json::Value,
    pub result_status: AuditStatus,
    pub error: Option<String>,
    pub duration_ms: i64,
    pub session_id: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
pub(crate) enum AuditStatus {
    Success,
    WhitelistDenied,
    ExecutionFailed,
}

pub(crate) struct AuditLogger {
    file: Mutex<Option<std::fs::File>>,
    session_id: String,
}

impl AuditLogger {
    pub fn new(audit_file: Option<&Path>) -> anyhow::Result<Self> {
        let session_id = uuid::Uuid::new_v4().to_string();
        let file = audit_file
            .map(|path| {
                std::fs::OpenOptions::new()
                    .create(true)
                    .append(true)
                    .open(path)
            })
            .transpose()?;
        Ok(Self {
            file: Mutex::new(file),
            session_id,
        })
    }

    pub fn session_id(&self) -> &str {
        &self.session_id
    }

    pub fn log(&self, entry: AuditEntry) {
        let line = serde_json::to_string(&entry).unwrap_or_else(|_| {
            r#"{"error":"audit serialization failed"}"#.to_string()
        });
        if let Ok(mut guard) = self.file.lock() {
            if let Some(ref mut file) = *guard {
                use std::io::Write;
                let _ = writeln!(file, "{line}");
            }
        }
        // 始终写 stderr 在开发阶段可见
        eprintln!("[audit] {line}");
    }
}
